core/src/main/java/com/graphhopper/routing/ch/CHAlgoFactoryDecorator.java

```diff
@@ -251,46 +252,22 @@ public RoutingAlgorithmFactory getDecoratedAlgorithmFactory(RoutingAlgorithmFact
+        Integer uTurnCosts = map.has(Routing.U_TURN_COSTS) ? map.getInt(Routing.U_TURN_COSTS, INFINITE_U_TURN_COSTS) : null;
+        CHProfile selectedProfile = selectProfile(map, edgeBased, uTurnCosts);
         for (PrepareContractionHierarchies p : getPreparations()) {
-            boolean weightingMatches = p.getCHProfile().getWeighting().matches(map);
-            if (weightingMatches) {
-                if (p.isEdgeBased()) {
-                    edgeBasedPCH = p;
-                } else {
-                    nodeBasedPCH = p;
-                }
+            if (p.getCHProfile().equals(selectedProfile)) {
+                return p;
             }
-            entriesStrs.add(p.getCHProfile().getWeighting() + "|" + (p.getCHProfile().isEdgeBased() ? "edge" : "node"));
-        }
-
-        if (edgeBasedPCH == null && nodeBasedPCH == null) {
-            throw new IllegalArgumentException("Cannot find CH RoutingAlgorithmFactory for weighting map " + map + " in entries: " + entriesStrs + ".");
         }
-        if (map.has(Parameters.Routing.EDGE_BASED)) {
-            boolean edgeBased = map.getBool(Parameters.Routing.EDGE_BASED, false);
-            if (edgeBased && edgeBasedPCH != null) {
-                return edgeBasedPCH;
-            }
-            if (!edgeBased && nodeBasedPCH != null) {
-                return nodeBasedPCH;
-            }
+        throw new IllegalStateException("Could not find CH preparation for selected profile: " + selectedProfile);
+    }
 
-            if (edgeBased) {
-                throw new IllegalArgumentException("Found a node-based CH preparation for weighting map " + map + ", but requested edge-based CH. " +
-                        "You either need to configure edge-based CH preparation or set the '" + Parameters.Routing.EDGE_BASED + "' " +
-                        "request parameter to 'false' (was 'true'). all entries: " + entriesStrs);
-            } else {
-                throw new IllegalArgumentException("Found an edge-based CH preparation for weighting map " + map + ", but requested node-based CH. " +
-                        "You either need to configure node-based CH preparation or set the '" + Parameters.Routing.EDGE_BASED + "' " +
-                        "request parameter to 'true' (was 'false'). all entries: " + entriesStrs);
-            }
-        } else {
-            // no edge_based parameter was set, we determine the CH preparation based on what is there (and prefer edge-based
-            // if we can choose)
-            return edgeBasedPCH != null ? edgeBasedPCH : nodeBasedPCH;
+    private CHProfile selectProfile(HintsMap map, Boolean edgeBased, Integer uTurnCosts) {
+        try {
+            return CHProfileSelector.select(chProfiles, map, edgeBased, uTurnCosts);
+        } catch (CHProfileSelectionException e) {
+            throw new IllegalArgumentException(e.getMessage());
```

Comment:
Is the checked exception CHProfileSelectionException required? Couldn't we let it extend DetailedIllegalArgumentException and avoid this method here and the 'throws CHProfileSelectionException' in a few methods? Or do you want it to be explicit? (I try to avoid checked exception as they are not often very helpful IMO)

Reply:
Ok, I changed it to a runtime exception. I do not necessarily think checked exceptions are never helpful because they give you a compile time check that helps you remembering to handle the error cases (for example here I think this is more of an advantage than it is a disadvantage that some 'intermediate' methods need to declare the exception). Anyway, either way is fine for me :+1:.